src/wallet.js

```diff
@@ -0,0 +1,273 @@
+/* @flow */
+import { HDNode } from 'bitcoinjs-lib'
+import { ecPairToHexString } from './utils'
+import crypto from 'crypto'
+
+const APPS_NODE_INDEX = 0
+const IDENTITY_KEYCHAIN = 888
+const BLOCKSTACK_ON_BITCOIN = 0
+
+const BITCOIN_BIP_44_PURPOSE = 44
+const BITCOIN_COIN_TYPE = 0
+const BITCOIN_ACCOUNT_INDEX = 0
+
+const EXTERNAL_ADDRESS = 'EXTERNAL_ADDRESS'
+const CHANGE_ADDRESS = 'CHANGE_ADDRESS'
+
+export type IdentityKeyPair = {
+  key: string,
+  keyID: string,
+  address: string,
+  appsNodeKey: string,
+  salt: string
+}
+
+function hashCode(string) {
+  let hash = 0
+  if (string.length === 0) return hash
+  for (let i = 0; i < string.length; i++) {
+    const character = string.charCodeAt(i)
+    hash = (hash << 5) - hash + character
+    hash = hash & hash
+  }
+  return hash & 0x7fffffff
+}
+
+function getNodePrivateKey(hdNode): string {
+  return ecPairToHexString(hdNode.keyPair)
+}
+
+function getNodePublicKey(hdNode): string {
+  return hdNode.keyPair.getPublicKeyBuffer().toString('hex')
+}
+
+/**
+ * The BlockstackWallet class manages the hierarchical derivation
+ *  paths for a standard blockstack client wallet. This includes paths
+ *  for bitcoin payment address, blockstack identity addresses, blockstack
+ *  application specific addresses.
+ */
+export class BlockstackWallet {
+  rootNode: HDNode
+
+  constructor(rootNode: HDNode) {
+    this.rootNode = rootNode
+  }
+
+  /**
+   * Initialize a blockstack wallet
+   * @param {Buffer} seed - the input seed for initializing the root node
+   *  of the hierarchical wallet
+   * @return {BlockstackWallet} the constructed wallet
+   */
+  static fromSeedBuffer(seed: Buffer): BlockstackWallet {
+    return new BlockstackWallet(HDNode.fromSeedBuffer(seed))
+  }
+
+  /**
+   * Initialize a blockstack wallet
+   * @param {string} keychain - the Base58 string used to initialize
+   *  the root node of the hierarchical wallet
+   * @return {BlockstackWallet} the constructed wallet
+   */
+  static fromBase58(keychain: string): BlockstackWallet {
+    return new BlockstackWallet(HDNode.fromBase58(keychain))
+  }
+
+  getIdentityPrivateKeychain(): HDNode {
+    return this.rootNode
+      .deriveHardened(IDENTITY_KEYCHAIN)
+      .deriveHardened(BLOCKSTACK_ON_BITCOIN)
+  }
+
+  getBitcoinPrivateKeychain(): HDNode {
+    return this.rootNode
+      .deriveHardened(BITCOIN_BIP_44_PURPOSE)
+      .deriveHardened(BITCOIN_COIN_TYPE)
+      .deriveHardened(BITCOIN_ACCOUNT_INDEX)
+  }
+
+  getBitcoinNode(addressIndex: number, chainType: string = EXTERNAL_ADDRESS): HDNode {
+    return BlockstackWallet.getNodeFromBitcoinKeychain(
+      this.getBitcoinPrivateKeychain().toBase58(),
+      addressIndex, chainType)
+  }
+
+  getIdentityAddressNode(identityIndex: number): HDNode {
+    const identityPrivateKeychain = this.getIdentityPrivateKeychain()
+    return identityPrivateKeychain.deriveHardened(identityIndex)
+  }
```

Comment:
This function doesn't seem to line up with `blockstack-browser`'s implementation, https://github.com/blockstack/blockstack-browser/blob/master/app/js/utils/account-utils.js#L212. Let me know if there's something I'm missing here.